substrafl/algorithms/pytorch/torch_fed_pca_algo.py

```diff
@@ -0,0 +1,304 @@
+        self.eigen_vectors = torch.nn.Linear(self.in_features, self.out_features, bias=False).to(self.device)
+        self.eigen_vectors.weight.data.to(self.device)
+
+    def forward(self, x):
+        """Perform dimensionality reduction.
+
+        Args:
+            x (torch.Tensor): inputs to map to reduced dim space.
+                shape of x is (B, D_IN)
+
+        Returns:
+            torch.Tensor: reduced dim vectors
+        """
+        eigen_vectors = self.eigen_vectors.weight.data
+        coefs = x @ eigen_vectors.T
+        return coefs @ eigen_vectors
+
+
+class TorchFedPCAAlgo(TorchAlgo):
+    """TorchFedPCAAlgo class, inheriting from TorchAlgo and designed to perform Principal Component Analysis (PCA).
+
+    Args:
+        dataset (torch.utils.data.Dataset): input data on which to perform PCA
+        in_features (int): input data dimensionality
+        out_features (int): number of dimensions to keep after PCA
+        batch_size (Optional[int]): mini-batch size
+        seed (int): random generator seed. Default to 1.
+        use_gpu (bool): whether to use GPU or not
+    """
+
+    def __init__(
+        self,
+        dataset: torch.utils.data.Dataset,
+        in_features: int,
+        out_features: int,
+        batch_size: Optional[int] = None,
+        seed: int = 1,
+        use_gpu: bool = True,
+        *args,
+        **kwargs,
+    ):
+        self.in_features = in_features
+        self.out_features = out_features
+        self._batch_size = batch_size
+        self.local_mean = None
+        self.local_covmat = None
+        self.round_counter = 0
+        self._seed = seed
+        torch.manual_seed(self._seed)
+
+        super().__init__(
+            model=TorchLinearModel(
+                self.in_features,
+                self.out_features,
+                device=self._get_torch_device(use_gpu),
+            ),
+            criterion=None,
+            optimizer=None,
+            index_generator=None,
+            dataset=dataset,
+            scheduler=None,
+            seed=self._seed,
+            use_gpu=use_gpu,
+            *args,
+            **kwargs,
+        )
+
+    def _local_predict(self, predict_dataset: torch.utils.data.Dataset, predictions_path: Path):
+        """Map inputs to lower dimensional space.
+
+        Args:
+            predict_dataset (torch.utils.data.Dataset): predict_dataset build from the x returned by the opener.
+            predictions_path (os.PathLike): path where to save predictions.
+
+        Important:
+            This function shall not be used before the second round to Fed PCA algo is
+            completed. Before that, the covariance matrix is not built.
+
+        Raises:
+            ValueError if the function is called before second round to Fed PCA algo is
+            completed.
+        """
+        if self.round_counter <= 2:
+            logger.warning(f"Evaluation ignored at round zero and one for {self.name} (pre-processing rounds).")
```

Comment:
The round concept shouldn't be in the algo. A better solution is to update the `predict` method within the strategy, that have the round_idx as parameter. Cf [the SingleOrgStrategy](https://github.com/Substra/substrafl/blob/b1fea19600059f5a00f7e102e9b6787ddbccf79d/substrafl/strategies/single_organization.py#L96) that face the same issue that you try to solve here.

Doing this, you can remove all references to `self.round_counter` in the algo.